src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,113 @@
 package core.basesyntax;
 
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_INDEX = 1.5;
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elements.length) {
+            arrayIncrease();
+        }
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        ensureCapacity(size + 1);
+        if (size == elements.length) {
+            arrayIncrease();
+        }
+
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds");
+        }
 
+        for (int i = size; i > index; i--) {
+            elements[i] = elements[i - 1];
+        }
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        ensureCapacity(size + list.size());
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+    }
 
+    private void ensureCapacity(int minCapacity) {
+        if (minCapacity > elements.length) {
+            int newCapacity = Math.max(elements.length + (elements.length >> 1), minCapacity);
+            Object[] newElements = new Object[newCapacity];
+            System.arraycopy(elements, 0, newElements, 0, size);
+            elements = newElements;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
+        checkIndex(index);
+        elements[index] = value;
+    }
 
+    private void arrayIncrease() {
+        int newCapacity = (int) (elements.length * GROWTH_INDEX);
+        ensureCapacity(newCapacity);
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final T removedElement = (T) elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[size - 1] = null;
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((elements[i] == null && element == null)
+                    || (elements[i] != null && elements[i].equals(element))) {
+                final T removedElement = (T) elements[i];
+                System.arraycopy(elements, i + 1, elements, i, size - i - 1);
+                elements[size - 1] = null;
+                size--;
+                return removedElement;
+            }
+        }
+        throw new java.util.NoSuchElementException("Element not found");
```

Comment:
You have created your own core.basesyntax.NoSuchElementException but you don't use it. why?

Reply:
[ERROR] Errors: 
[ERROR]   ArrayListTest.removeElementFromArrayListByNonExistentValue ┬╗  Unexpected exception, expected<java.util.NoSuchElementException> but was<core.basesyntax.ArrayListIndexOutOfBoundsException>
Maven wants me to use java.util.NoSuchElementException instead of mine  ArrayListIndexOutOfBoundsException
because the condition of task says to throw java.util.NoSuchElementException, but not to create it.

Reply:
Maven wanted java.util.NoSuchElementException instead of my custom but I forgot to delete it